Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -0,0 +1,361 @@
+  rfl
+
+@[simp]
+theorem toNimber_min (a b : Ordinal) :
+    toNimber (min a b) = min (toNimber a) (toNimber b) :=
+  rfl
+
+end Ordinal
+
+/-! ### Nimber addition -/
+
+namespace Nimber
+
+variable {a b c : Nimber.{u}}
+
+/-- Nimber addition is recursively defined so that `a + b` is the smallest number not equal to
+`a' + b` or `a + b'` for `a' < a` and `b' < b`. -/
+-- We write the binders like this so that the termination checker works.
+protected def add (a b : Nimber.{u}) : Nimber.{u} :=
+  sInf {x | (∃ a', ∃ (_ : a' < a), Nimber.add a' b = x) ∨
+    ∃ b', ∃ (_ : b' < b), Nimber.add a b' = x}ᶜ
+termination_by (a, b)
+
+instance : Add Nimber :=
+  ⟨Nimber.add⟩
+
+theorem add_def (a b : Nimber) :
+    a + b = sInf {x | (∃ a' < a, a' + b = x) ∨ ∃ b' < b, a + b' = x}ᶜ := by
+  change Nimber.add a b = _
+  rw [Nimber.add]
+  simp_rw [exists_prop]
+  rfl
+
+/-- The set in the definition of `Nimber.add` is nonempty. -/
+theorem add_nonempty (a b : Nimber) :
+    {x | (∃ a' < a, a' + b = x) ∨ ∃ b' < b, a + b' = x}ᶜ.Nonempty := by
+  use Ordinal.blsub₂ (succ a) (succ b) @fun x _ y _ => Nimber.add x y
+  simp only [Set.mem_compl_iff, Set.mem_setOf_eq, not_or, not_exists, not_and]
+  constructor <;>
+  intro x hx <;>
+  apply ne_of_lt <;>
+  apply Ordinal.lt_blsub₂
+  exacts [hx.trans <| lt_succ _, lt_succ _, lt_succ _, hx.trans <| lt_succ _]
+
+theorem exists_of_lt_add (h : c < a + b) : (∃ a' < a, a' + b = c) ∨ ∃ b' < b, a + b' = c := by
+  rw [add_def] at h
+  have := not_mem_of_lt_csInf h ⟨_, bot_mem_lowerBounds _⟩
+  rwa [Set.mem_compl_iff, not_not] at this
+
+theorem add_le_of_forall_ne (h₁ : ∀ a' < a, a' + b ≠ c) (h₂ : ∀ b' < b, a + b' ≠ c) :
+    a + b ≤ c := by
+  by_contra! h
+  have := exists_of_lt_add h
+  tauto
+
+private theorem add_ne_of_lt (a b : Nimber) :
+    (∀ a' < a, a' + b ≠ a + b) ∧ ∀ b' < b, a + b' ≠ a + b := by
+  have H := csInf_mem (add_nonempty a b)
+  rw [← add_def] at H
+  simpa using H
+
+instance : IsLeftCancelAdd Nimber := by
+  constructor
+  intro a b c h
+  apply le_antisymm <;>
+  apply le_of_not_lt
+  · exact fun hc => (add_ne_of_lt a b).2 c hc h.symm
+  · exact fun hb => (add_ne_of_lt a c).2 b hb h
+
+instance : IsRightCancelAdd Nimber := by
+  constructor
+  intro a b c h
+  apply le_antisymm <;>
+  apply le_of_not_lt
+  · exact fun hc => (add_ne_of_lt a b).1 c hc h.symm
+  · exact fun ha => (add_ne_of_lt c b).1 a ha h
+
+-- Ideally the proof would be an easy induction on `add_def`, but rewriting under binders trips up
+-- the termination checker.
+protected theorem add_comm (a b : Nimber) : a + b = b + a := by
+  apply le_antisymm <;>
+  apply add_le_of_forall_ne <;>
+  intro x hx
+  on_goal 1 => rw [Nimber.add_comm x, add_ne_add_right]
+  on_goal 2 => rw [Nimber.add_comm a, add_ne_add_left]
+  on_goal 3 => rw [← Nimber.add_comm a, add_ne_add_right]
+  on_goal 4 => rw [← Nimber.add_comm x, add_ne_add_left]
+  all_goals exact hx.ne
+termination_by (a, b)
+
+@[simp]
```

Comment:
I don't think this should be a simp lemma. For comparison, `sub_eq_zero` isn't `@[simp]`.